src/web100srv.c

```diff
@@ -1541,70 +1473,67 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
+        log_println(0, "Cannot create worker thread for writing cpu usage!");
         workerThreadId = 0;
         memset(cputimelog, 0, 256);
       }
     }
   }
+
   // write the incoming connection data into the log file
   fp = fopen(get_logfile(), "a");
   if (fp == NULL) {
-    log_println(0, "Unable to open log file '%s', continuing on "
-                "without logging", get_logfile());
+    log_println(0,
+                "Unable to open log file '%s', continuing on without logging",
+                get_logfile());
   } else {
-    fprintf(fp, "%15.15s  %s port %d\n", ctime(&tt)+4, rmt_host,
-            testPort);
+    fprintf(fp, "%15.15s  %s port %d\n", ctime(&tt) + 4, rmt_host, testPort);
     if (cputime && workerThreadId) {
       log_println(1, "cputime trace file: %s\n", cputimelog);
       fprintf(fp, "cputime trace file: %s\n", cputimelog);
     }
     fclose(fp);
   }
-  close(chld_pipe[0]);
+  close(parent_pipe);
 
   // Set test options
-  if (t_opts & TEST_MID)
-    testopt.midopt = TOPT_ENABLED;
-  if (t_opts & TEST_SFW)
-    testopt.sfwopt = TOPT_ENABLED;
-  if (t_opts & TEST_META)
-    testopt.metaopt = TOPT_ENABLED;
-  if (t_opts & TEST_C2S)
-    testopt.c2sopt = TOPT_ENABLED;
-  if (t_opts & TEST_S2C)
-    testopt.s2copt = TOPT_ENABLED;
+  if (t_opts & TEST_MID) testopt.midopt = TOPT_ENABLED;
+  if (t_opts & TEST_SFW) testopt.sfwopt = TOPT_ENABLED;
+  if (t_opts & TEST_META) testopt.metaopt = TOPT_ENABLED;
+  if (t_opts & TEST_C2S) testopt.c2sopt = TOPT_ENABLED;
+  if (t_opts & TEST_S2C) testopt.s2copt = TOPT_ENABLED;
   // die in 120 seconds, but only if a test doesn't get started
   alarm(120);
   // reset alarm() before every test
-  log_println(6, "setting master alarm() to 120 seconds, tests "
-              "must start (complete?) before this timer expires");
+  log_println(6,
+              "setting master alarm() to 120 seconds, tests must start "
+              "(complete?) before this timer expires");
 
   // run tests based on options
   if (strncmp(test_suite, "Invalid", 7) != 0) {
-    log_println(3, "Valid test sequence requested, run test for "
-                "client=%d", getpid());
-    retcode = run_test(agent, ctlsockfd, &testopt, test_suite);
+    log_println(3, "Valid test sequence requested, run test for client=%d",
+                getpid());
+    retcode = run_test(agent, &ctl, &testopt, test_suite, ssl_context);
   }
 
   // conclude all test runs
   if (retcode == 0) {
     log_println(3, "Successfully returned from run_test() routine");
   } else {
-    log_println(3, "Child %d returned non-zero (%d) from "
-                "run_test() results some test failed!", getpid(),
-                retcode);
+    log_println(3,
+                "Child %d returned non-zero (%d) from "
+                "run_test() results some test failed!",
+                getpid(), retcode);
     child_sig(0);
   }
-  close(ctlsockfd);
+  SSL_free(ctl.ssl);
+  close(ctl.socket);
```

Comment:
Can we add an `int shutdown` (like a boolean) param to close_connection so that we have a consistent way of freeing connection resources and zeroing them out even if we don't want to disconnect?